File: Basic/Basic_Recursion/valid_palindrome.py
def helper(s, low, high):
    if low > high:
        return True
    if not s[low].isalnum():
        return helper(s, low + 1, high)
    if not s[high].isalnum():
        return helper(s, low, high - 1)
    if s[low].lower() != s[high].lower():
        return False
    return helper(s, low + 1, high - 1)


def is_palindrome(s):
    return helper(s, 0, len(s) - 1)

File: Basic/Basic_Recursion/ValidPalindrome.cpp
#include <cctype>
#include <string>

class ValidPalindrome {
public:
    static bool isPalindrome(const std::string& s) {
        int left = 0, right = static_cast<int>(s.size()) - 1;
        while (left < right) {
            while (left < right && !std::isalnum(static_cast<unsigned char>(s[left]))) ++left;
            while (left < right && !std::isalnum(static_cast<unsigned char>(s[right]))) --right;
            if (std::tolower(static_cast<unsigned char>(s[left])) !=
                std::tolower(static_cast<unsigned char>(s[right]))) return false;
            ++left;
            --right;
        }
        return true;
    }
};